packages/beacon-node/src/chain/blocks/blockInput/blockInput.ts

```diff
@@ -0,0 +1,875 @@
+    }
+  | {
+      blockStatus: BlockStatus.HasBlock;
+      daStatus: DAStatus.CompleteData;
+      block: SignedBeaconBlock<F>;
+      source: SourceMeta;
+      timeCompleteSec: number;
+    };
+
+abstract class AbstractBlockInput<F extends ForkName = ForkName, TData extends DAData = DAData>
+  implements IBlockInput<F, TData>
+{
+  abstract type: DAType;
+  daRequirement: DARequirement;
+  timeBeginSec: number;
+
+  forkName: ForkName;
+  slot: Slot;
+  blockRootHex: string;
+  parentRootHex: string;
+
+  abstract state: BlockInputState<F>;
+
+  protected blockPromise = createPromise<SignedBeaconBlock<F>>();
+  protected dataPromise = createPromise<TData>();
+  protected bothPromise = createPromise<this>();
+
+  constructor(init: BlockInputInit) {
+    this.daRequirement = init.daRequirement;
+    this.timeBeginSec = init.timeBeginSec;
+    this.forkName = init.forkName;
+    this.slot = init.slot;
+    this.blockRootHex = init.blockRootHex;
+    this.parentRootHex = init.parentRootHex;
+  }
+
+  abstract addBlock(props: AddBlock<F>): void;
+
+  hasBlock(): boolean {
+    return this.state.blockStatus === BlockStatus.HasBlock;
+  }
+
+  getBlock(): SignedBeaconBlock<F> {
+    if (this.state.blockStatus !== BlockStatus.HasBlock) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISSING_BLOCK,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot getBlock from BlockInput without a block"
+      );
+    }
+    return this.state.block;
+  }
+
+  getBlockSource(): SourceMeta {
+    if (this.state.blockStatus !== BlockStatus.HasBlock) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISSING_BLOCK,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot getBlockSource from BlockInput without a block"
+      );
+    }
+    return this.state.source;
+  }
+
+  hasData(): boolean {
+    return this.state.daStatus === DAStatus.CompleteData;
+  }
+
+  hasBlockAndData(): boolean {
+    return this.state.blockStatus === BlockStatus.HasBlock && this.state.daStatus === DAStatus.CompleteData;
+  }
+
+  getLogMeta(): LogMetaBasic {
+    return {
+      blockRoot: prettyBytes(this.blockRootHex),
+      slot: this.slot,
+    };
+  }
+
+  getTimeComplete(): number {
+    if (this.state.blockStatus !== BlockStatus.HasBlock || this.state.daStatus !== DAStatus.CompleteData) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISSING_TIME_COMPLETE,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot getTimeComplete from BlockInput without a block and data"
+      );
+    }
+    return this.state.timeCompleteSec;
+  }
+
+  async waitForBlock(timeout: number, signal?: AbortSignal): Promise<SignedBeaconBlock<F>> {
+    if (this.state.blockStatus === BlockStatus.MissingBlock) {
+      return await withTimeout(() => this.blockPromise.promise, timeout, signal);
+    }
+    return this.state.block;
+  }
```

Comment:
Can we just return `withTimeout` directly.  Will have the same effect.

```suggestion
 waitForBlock(timeout: number, signal?: AbortSignal): Promise<SignedBeaconBlock<F>> {
      return withTimeout(() => this.blockPromise.promise, timeout, signal);
  }
```

Reply:
same for `waitForData`

Reply:
done